contracts/bns/BNSRegistrar.sol

```diff
@@ -0,0 +1,133 @@
+        bytes4(keccak256("supportsInterface(bytes4)"));
+    bytes4 private constant ERC721_ID =
+        bytes4(
+            keccak256("balanceOf(address)") ^
+                keccak256("ownerOf(uint256)") ^
+                keccak256("approve(address,uint256)") ^
+                keccak256("getApproved(uint256)") ^
+                keccak256("setApprovalForAll(address,bool)") ^
+                keccak256("isApprovedForAll(address,address)") ^
+                keccak256("transferFrom(address,address,uint256)") ^
+                keccak256("safeTransferFrom(address,address,uint256)") ^
+                keccak256("safeTransferFrom(address,address,uint256,bytes)")
+        );
+    bytes4 private constant RECLAIM_ID =
+        bytes4(keccak256("reclaim(uint256,address)"));
+
+    constructor(
+        BNS _bns,
+        bytes32 _baseNode,
+        ReverseRegistrar _reverseRegistrar
+    ) ERC721("", "") {
+        bns = _bns;
+        baseNode = _baseNode;
+        reverseRegistrar = _reverseRegistrar;
+    }
+
+    modifier live() {
+        require(bns.owner(baseNode) == address(this));
+        _;
+    }
+
+    // Set the resolver for the TLD this registrar manages.
+    function setResolver(address resolver) external override onlyOwner {
+        bns.setResolver(baseNode, resolver);
+    }
+
+    /**
+     * @dev Register a name.
+     * @param name The label to be registered and used for ID (keccak256(name)).
+     * @param owner The address that should own the registration.
+     * @param resolver The address of the resolver.
+     */
+    function register(
+        string calldata name,
+        address owner,
+        address resolver
+    ) external override {
+        _register(name, owner, resolver, true, true);
+    }
+
+    /**
+     * @dev Register a name, without modifying the registry.
+     * @param name The label to be registered and used for ID (keccak256(name))
+     * @param owner The address that should own the registration.
+     * @param resolver The address of the resolver.
+     */
+    function registerOnly(
+        string calldata name,
+        address owner,
+        address resolver
+    ) external {
+        _register(name, owner, resolver, false, false);
+    }
+
+    function _register(
+        string memory name,
+        address owner,
+        address resolver,
+        bool updateRegistry,
+        bool reverseRecord
+    ) internal live {
+        uint256 id = uint256(keccak256(abi.encodePacked(name)));
+        _mint(owner, id);
+        if (updateRegistry) {
+            bns.setSubnodeOwner(baseNode, bytes32(id), owner);
+        }
+        if (reverseRecord) {
+            _setReverseRecord(name, resolver, msg.sender);
+        }
+
+        emit NameRegistered(id, owner);
+    }
+
+    /**
+     * @dev Reclaim ownership of a name in BNS, if you own it in the registrar.
+     */
+    function reclaim(uint256 id, address owner) external override live {
+        require(_isApprovedOrOwner(msg.sender, id));
```

Comment:
One thing we may want to discuss/think about is having an "admin" wallet as well, so that we can reclaim addresses if they're lost and bring them back to the registry. It's a bit more "centralized", but maybe worth having? IDK.